Contributions/Linux_Memory_Management_Essentials.md

```diff
@@ -0,0 +1,249 @@
+  the one used while writing the document (6.9).
+* When referring to specific HW features, the document refers to the ARM64 architecture.
+
+## **Purpose of the document**
+This document aims to provide an holistic view of what happens in Linux memory management, so that
+one is at least aware of certain features and can use this document as a jumping pad toward more detailed documentation.
+Or even to the code base.
+
+## **Structure of the document**
+The document is divided in two parts, based on the destinatary of the memory allocations discussed: kernel-space or user-space.
+Individual points are numbered for ease of reference, but the numbering is not meant to represent any sequence.
+
+## **Memory management in Linux**
+
+### **Kernel-space memory allocations**
+
+#### **Directly Verifiable Assertions**
+The following section presents a set of statements that can be objectively verified e.g. by inspecting the sources.
+
+1. unlike processes memory, kernel memory pages are not swapped, nor dropped silently by the kernel itself,
+   although an hypervisor will do to a VM what the kernel does to a process (but this is beyond the control of the kernel)
+2. the kernel context (usually EL1 on ARM64) uses one single memory map (page tables) across all the cores
+   executing in kernel mode
+3. on 64 bit systems (e.g. ARM64 and x86_64), usually almost all physical memory is mapped in a
+   (semi)contiguous (there can be holes) range. Memory within this range is both virtually and physically contiguous.
+4. physically contiguous memory is treated as a scarce resource, and typically is not provided to userspace,
+   unless it explicitly asks for it (e.g. for implementing some DMA buffer)
+5. the kernel can access userspace memory in three ways:
+   1. through the userspace mapping 
+      1. this type of access is limited to few functions, like copy_to_user()/copy_from_user() and put_user()/get_user()
+      2. outside the execution of the functions mentioned, this type of access is not possible, because the userspace mappings are made available only while executing such functions
+      3. the userspace memory map can implement HW protections against being misused by the kernel
+      4. the kernel is able to access process pages in the same sequence and with the same mappings as the process does
+   2. through a memory buffer (e.g. a memory area where the kernel regularly needs to perform large amount of read/write operations, like a network buffer)
+      1. the memory used by a buffer is well defined and delimited, rather than a generic area, and it is specifically reserved for this purpose
+      2. the kernel uses own mappings in EL1, while the user space uses the EL0 ones
+      3. very often the kernel doesn't actually access this area directly, but it rather configures a DMA controller to do the transfers, directly to the physical memory.
```

Comment:
Why are you thinking of DMA buffers? You could also have non-dma-able  phisical memory mapped both to user virtual address space as well as to Kernel virtual address space. I would consider the hazard associated with dma buffers as a separate additional point

Reply:
"very often" doesn't mean "always"; besides, the scenario where the kernel writes directly already exists, and it is done by using the user space mappings through TTBR0.
The DMA is relevant because it completely bypasses the MMU

Reply:
This is a valid qualification, but perhaps it could be reworded like this:

```suggestion
      3. in many cases, the kernel does not access this area directly, but rather configures a DMA controller to complete the transfer directly to the physical memory (i.e. bypassing the MMU)
```